design/src/rfcs/rfc0022_refine_builder.md

```diff
@@ -0,0 +1,556 @@
+
+### Cut them down: going from 2N+1 to 2 generic parameters
+
+The previous two examples should have convinced you that the 2N+1 generic parameters on `PokemonServiceBuilder` harm the ergonomics of our API.  
+Can we get rid of them?
+
+Yes! Let's look at one possible approach:
+
+```rust
+pub struct PokemonServiceBuilder<Body, Plugin> {
+    check_health: Option<Route<Body>>,
+    do_nothing: Option<Route<Body>>,
+    get_pokemon_species: Option<Route<Body>>,
+    get_server_statistics: Option<Route<Body>>,
+    capture_pokemon: Option<Route<Body>>,
+    get_storage: Option<Route<Body>>,
+    plugin: Plugin,
+}
+```
+
+We no longer store the raw handlers inside `PokemonServiceBuilder`.
+We eagerly upgrade the operation handlers to a `Route` instance when they are registered with the builder. The existing API performs the upgrade when `build` is called, forcing `PokemonServiceBuilder` to store the raw handlers and keep two generic parameters around (`OpX` and `ExtsX`) for each operation.  
+
+There is one downside to this alternative approach - all plugins must be specified upfront, when creating an instance of the builder.
+Plugins cannot be modified through `PokemonServiceBuilder`'s API if we want to guarantee that all operation handlers are upgraded to a `Route` using the same plugins.
+
+```rust
+impl PokemonService<()> {
+    /// Constructs a builder for [`PokemonService`].
+    pub fn builder<Body, Plugin>(plugin: Plugin) -> PokemonServiceBuilder<Body, Plugin> {
+        PokemonServiceBuilder {
+            check_health: None,
+            do_nothing: None,
+            get_pokemon_species: None,
+            get_server_statistics: None,
+            capture_pokemon: None,
+            get_storage: None,
+            plugin,
+        }
+    }
+}
+```
+
+In practice, this is unlikely to have any impact on developers using `smithy-rs`.  
+We have seen how cumbersome it is to break the startup logic into different functions using the current service builder API. Developers are most likely specifying all plugins and routes in the same function even if the current API allows them to intersperse route registrations and plugin registrations: they would simply have to re-order their registration statements to adopt the API proposed in this RFC.
+
+### Alternatives: boxed trait objects 
+
+Trait objects are another common approach to achieve type erasure: instead of storing the raw handler in `PokemonServiceBuilder`, we box it and cast it to a `dyn Upgradable<...>`.  
+Using trait objects would allow us to delay upgrading to `Route` until `build` is called, mimicking the behaviour of the current API and allowing developers to register plugins at any point before calling `build`.  
+Using trait objects would unfortunately have a non-zero runtime impact: we would be introducing an extra layer of indirection (double-boxing the handler), affecting the performance of our application at runtime.  
```

Comment:
I think this is worth exploring. Following a vtable has negligible performance impact for our use cases. If you can flatten the type that the router ends up holding to just a few boxes, it's likely ok.

Reply:
I've significantly expanded this section exploring two different approaches for lazy type erasure. 
I don't believe it's actually practical moving forward with either of them - they do improve slightly on the current status quo but they carry a significant complexity budget in terms of implementation and ongoing maintenance.
We spent some time we @hlbarber to play around with ideas and they get quite complex quite quickly.